### PR TITLE
Security: Make CandidateSet timeout and initial fanout more reliable

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

1. Zebra applies a separate timeout to each fanout readiness check, resulting in repeated timeouts for slow peer sets.

2. Zebra uses the number of resolved peer addresses for its initial update fanout.

So it was possible for all peers to fail, and for Zebra to hang on the first `CandidateSet::update`. Also, Zebra could send a fanout for each initial peer, regardless of whether that peer's connection was successful.

This is part of PR #2160, which implements part of the #1849 refactor needed for #1868.

## Solution

Security
- Apply a timeout to the entire CandidateSet::update
- Stop using very large fanout limits during initialization

Fixes
- Reliability: Wait for at least one successful peer before trying an update
- Usability: Warn if there are no successful initial peers

Refactor
- Split CandidateSet::update into separate functions
- Update `MetaAddr::new_gossiped_meta_addr` to the latest API from #2160

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Integration Tests

## Review

@jvff should review and merge this PR before starting work on #1868

## Follow Up Work

This refactor should unblock #1868.

### CODE REVIEW

**Comment by teor2345** [comment]
> ### All checks have passed
> 16 successful checks

Well that's one part of #2160 we can safely merge.

File: zebra-network/src/meta_addr.rs [hunk 1]
```diff
@@ -137,17 +137,17 @@ pub struct MetaAddr {
...
 
 impl MetaAddr {
-    /// Create a new `MetaAddr` from the deserialized fields in a gossiped
-    /// peer `Addr` message.
-    pub fn new_gossiped(
-        addr: &SocketAddr,
-        services: &PeerServices,
-        last_seen: &DateTime<Utc>,
+    /// Create a new gossiped [`MetaAddr`], based on the deserialized fields from
+    /// a gossiped peer [`Addr`] message.
```
Comment (jvff) [inline zebra-network/src/meta_addr.rs:141]:
```suggestion
    /// a gossiped peer [`Addr`][crate::protocol::external::Message::Addr] message.
```

File: zebra-network/src/peer_set/candidate_set.rs [hunk 2]
```diff
@@ -133,17 +133,17 @@ where
 
     /// Update the peer set from the network, using the default fanout limit.
     ///
-    /// See `update_initial` for details.
+    /// See [`update_initial`] for details.
```
Comment (jvff) [inline zebra-network/src/peer_set/candidate_set.rs:136]:
```suggestion
    /// See [`update_initial`][Self::update_initial] for details.
```

File: zebra-network/src/peer_set/candidate_set.rs [hunk 3]
```diff
@@ -154,20 +154,48 @@ where
...
+    /// [`next`] puts peers into the [`AttemptPending`] state.
     pub async fn update_initial(&mut self, fanout_limit: usize) -> Result<(), BoxError> {
-        self.update_inner(Some(fanout_limit)).await
+        self.update_timeout(Some(fanout_limit)).await
+    }
+
+    /// Update the peer set from the network, limiting the fanout to
+    /// `fanout_limit`, and imposing a timeout on the entire fanout.
+    ///
+    /// See [`update_initial`] for details.
```
Comment (jvff) [inline zebra-network/src/peer_set/candidate_set.rs:169]:
```suggestion
    /// See [`update_initial`][Self::update_initial] for details.
```

File: zebra-network/src/peer_set/candidate_set.rs [hunk 4]
```diff
@@ -154,20 +154,48 @@ where
...
+
+        Ok(())
     }
 
     /// Update the peer set from the network, limiting the fanout to
     /// `fanout_limit`.
     ///
-    /// See `update_initial` for details.
-    async fn update_inner(&mut self, fanout_limit: Option<usize>) -> Result<(), BoxError> {
+    /// See [`update_initial`] for details.
```
Comment (jvff) [inline zebra-network/src/peer_set/candidate_set.rs:193]:
```suggestion
    /// See [`update_initial`][Self::update_initial] for details.
```

File: zebra-network/src/peer_set/candidate_set.rs [hunk 5]
```diff
@@ -154,20 +154,48 @@ where
     /// errors on permanent failures.
     ///
     /// The handshaker sets up the peer message receiver so it also sends a
-    /// `Responded` peer address update.
+    /// [`Responded`] peer address update.
     ///
-    /// `report_failed` puts peers into the `Failed` state.
+    /// [`report_failed`] puts peers into the [`Failed`] state.
```
Comment (jvff) [inline zebra-network/src/peer_set/candidate_set.rs:159]:
```suggestion
    /// [`report_failed`][Self::report_failed] puts peers into the [`Failed`] state.
```

File: zebra-network/src/peer_set/candidate_set.rs [hunk 6]
```diff
@@ -154,20 +154,48 @@ where
...
     ///
-    /// `next` puts peers into the `AttemptPending` state.
+    /// [`next`] puts peers into the [`AttemptPending`] state.
```
Comment (jvff) [inline zebra-network/src/peer_set/candidate_set.rs:161]:
```suggestion
    /// [`next`][Self::next] puts peers into the [`AttemptPending`] state.
    ///
    /// [`Responded`]: crate::PeerAddrState::Responded
    /// [`NeverAttemptedGossiped`]: crate::PeerAddrState::NeverAttemptedGossiped
    /// [`Failed`]: crate::PeerAddrState::Failed
    /// [`AttemptPending`]: crate::PeerAddrState::AttemptPending
```

File: zebra-network/src/peer_set/candidate_set.rs [hunk 7]
```diff
@@ -154,20 +154,48 @@ where
...
     /// Update the peer set from the network, limiting the fanout to
     /// `fanout_limit`.
     ///
-    /// See `update_initial` for details.
-    async fn update_inner(&mut self, fanout_limit: Option<usize>) -> Result<(), BoxError> {
+    /// See [`update_initial`] for details.
+    ///
+    /// # Correctness
+    ///
+    /// This function does not have a timeout. Use [`update_timeout`] instead.
```
Comment (jvff) [inline zebra-network/src/peer_set/candidate_set.rs:197]:
```suggestion
    /// This function does not have a timeout. Use [`update_timeout`][Self::update_timeout] instead.
```

**Review by jvff — APPROVED** [review]
I updated some documentation links, not sure if that's too nitpicky :/

Overall, looks good to me!

**Comment by teor2345** [comment]
> I updated some documentation links, not sure if that's too nitpicky :/

That's great, but I had auto-merge on, so the PR merged without your suggestions.

I'll go open another PR with those suggestions now.

**Comment by jvff** [comment]
Ah, sorry. I didn't know about that. Next time should I just leave the review as comments and only approve after all suggestions have been at least discussed?

**Comment by teor2345** [comment]
> Ah, sorry. I didn't know about that. Next time should I just leave the review as comments and only approve after all suggestions have been at least discussed?

Actually I think that was GitHub's fault? 🤷 

I wasn't expecting an auto-merge to merge when there are outstanding suggestions.

I'll avoid setting auto-merge on new PRs in future. (I might set it when everyone has agreed on the changes, and we're just waiting for them to pass CI.)

**Comment by teor2345** [comment]
Also GitHub should have an "open a new PR with these suggestions" feature.

**Pull request merged**